File: tomviz/python/tomviz/pipeline/sources/python_source.py
###############################################################################
# This source file is part of the Tomviz project, https://tomviz.org/.
# It is released under the 3-Clause BSD License, see "LICENSE".
###############################################################################
"""PythonSource — schema-v2 Python source node for the CLI runtime.

Mirrors the C++ ``tomviz::pipeline::PythonSource``. The user-facing class
is a subclass of :class:`tomviz.nodes.SourceNode` whose ``produce``
method returns the outputs dict."""

from tomviz.pipeline.node import SourceNode
from tomviz.pipeline.python_node_backend import PythonNodeBackend


class PythonSource(SourceNode):
    type_name = 'source.python'

    def __init__(self):
        super().__init__()
        self._backend = PythonNodeBackend()

    # ---- description / script -----------------------------------------

    def set_json_description(self, json_str: str):
        self._backend.set_json_description(json_str)
        if self._backend.default_label:
            self.label = self._backend.default_label
        # Sources have no inputs — pass a null AddInputFn. A v2 source
        # description that erroneously declares inputs is silently
        # dropped here; the menu-routing reaction in the app rejects
        # the same case at construction time.
        self._backend.apply_description(
            None,
            lambda name, ptype, persistent: self.add_output(
                name, ptype, persistent=persistent))

    @property
    def json_description(self) -> str:
        return self._backend.json_description

    @property
    def script(self) -> str:
        return self._backend.script

    @script.setter
    def script(self, value: str):
        self._backend.set_script(value)

    @property
    def parameters(self) -> dict:
        return self._backend.parameters

    # ---- serialize / deserialize --------------------------------------

    def deserialize(self, data: dict) -> bool:
        self._backend.apply_serialized_fields(
            data,
            None,
            lambda name, ptype, persistent: (
                self.add_output(name, ptype, persistent=persistent)
                if self.output_port(name) is None else None))
        return super().deserialize(data)

    # ---- execution ----------------------------------------------------

    def execute(self) -> bool:
        outputs = self._backend.run_source(self)
        if not outputs and self.output_ports():
            return False
        for name, port_data in outputs.items():
            port = self.output_port(name)
            if port is not None:
                port.set_data(port_data)
        return True

File: tomviz/pipeline/sources/PythonSource.cxx
/* This source file is part of the Tomviz project, https://tomviz.org/.
   It is released under the 3-Clause BSD License, see "LICENSE". */

#include "PythonSource.h"

#include "CustomNodeWidgetRegistry.h"
#include "CustomPythonNodeWidget.h"
#include "ExternalNodeExecutor.h"
#include "OutputPort.h"
#include "ParameterBindingUtils.h"
#include "PythonNodeEditorWidget.h"

namespace tomviz {
namespace pipeline {

PythonSource::PythonSource(QObject* parent) : SourceNode(parent)
{
}

void PythonSource::setJSONDescription(const QString& json)
{
  m_backend.setJSONDescription(json);
  if (!m_backend.defaultLabel().isEmpty()) {
    setLabel(m_backend.defaultLabel());
  }
  setSupportsCancel(m_backend.supportsCancel());
  setSupportsCompletion(m_backend.supportsComplete());
  // Shape mismatch (a transform-shaped description loaded by a source
  // shell) is rejected upstream by the menu-routing reaction; loading
  // a saved state file trusts its serialized type string. Either way,
  // a SourceNode never accepts inputs — silently ignore any inputs[]
  // entries the description still carries.
  m_backend.applyDescription(
    nullptr,
    [this](const QString& name, PortType type) {
      return addOutput(name, type);
    });
}

QString PythonSource::jsonDescription() const
{
  return m_backend.jsonDescription();
}

QStringList PythonSource::reconfigureDescription(const QString& json)
{
  auto reset = m_backend.reconfigure(json);
  setSupportsCancel(m_backend.supportsCancel());
  setSupportsCompletion(m_backend.supportsComplete());
  return reset;
}

void PythonSource::setScript(const QString& script)
{
  m_backend.setScript(script);
}

QString PythonSource::scriptSource() const
{
  return m_backend.scriptSource();
}

void PythonSource::setParameter(const QString& name, const QVariant& value)
{
  m_backend.setParameter(name, value);
}

QVariant PythonSource::parameter(const QString& name) const
{
  return m_backend.parameter(name);
}

QMap<QString, QVariant> PythonSource::parameters() const
{
  return m_backend.parameters();
}

QString PythonSource::operatorName() const
{
  return m_backend.operatorName();
}

bool PythonSource::hasPropertiesWidget() const
{
  return true;
}

EditNodeWidget* PythonSource::createPropertiesWidget(Pipeline* pipeline,
                                                     QWidget* parent)
{
  PythonNodeEditorWidget::CustomWidgetFactory factory;
  bool customNeedsData = false;
  auto widgetID = m_backend.customWidgetID();
  const CustomWidgetInfo* info =
    widgetID.isEmpty() ? nullptr : findCustomNodeWidget(widgetID);
  if (info && info->create) {
    // Sources have no inputs — needsData should be irrelevant here,
    // but honor the flag in case a registration unexpectedly sets it.
    customNeedsData = info->needsData;
    factory = [this, info](QWidget* p) -> CustomPythonNodeWidget* {
      auto* w = info->create(collectInputs(), p);
      if (w) {
        w->setScript(m_backend.scriptSource());
      }
      return w;
    };
  }

  QString currentType;
  QString currentEnvPath;
  if (auto* ext = qobject_cast<ExternalNodeExecutor*>(nodeExecutor())) {
    currentType = ExternalNodeExecutor::typeString();
    currentEnvPath = ext->envPath();
  }

  auto* widget = new PythonNodeEditorWidget(
    this, pipeline, label(), m_backend.scriptSource(),
    m_backend.jsonDescription(), m_backend.parameters(), currentType,
    currentEnvPath, factory, customNeedsData, parent);

  connect(widget, &PythonNodeEditorWidget::applied, this,
          [this](const PythonNodeEdits& edits) {
            bool changed = false;

            // Description first: it decides which parameters exist, so
            // the values below have to land on the new declarations.
            if (m_backend.jsonDescription() != edits.jsonDescription) {
              reconfigureDescription(edits.jsonDescription);
              changed = true;
            }

            if (label() != edits.label) {
              setLabel(edits.label);
              changed = true;
            }

            if (m_backend.scriptSource() != edits.script) {
              m_backend.setScript(edits.script);
              changed = true;
            }

            for (auto it = edits.values.constBegin();
                 it != edits.values.constEnd(); ++it) {
              if (m_backend.parameter(it.key()) != it.value()) {
                changed = true;
              }
              m_backend.setParameter(it.key(), it.value());
            }

            auto* currentExternal =
              qobject_cast<ExternalNodeExecutor*>(nodeExecutor());
            if (edits.executorType.isEmpty()) {
              if (nodeExecutor() != nullptr) {
                setNodeExecutor(nullptr);
                changed = true;
              }
            } else if (edits.executorType ==
                       ExternalNodeExecutor::typeString()) {
              if (currentExternal) {
                if (currentExternal->envPath() != edits.executorEnvPath) {
                  currentExternal->setEnvPath(edits.executorEnvPath);
                  changed = true;
                }
              } else {
                setNodeExecutor(
                  new ExternalNodeExecutor(edits.executorEnvPath));
                changed = true;
              }
            }

            if (changed) {
              emit parametersApplied();
            }
          });

  wireParameterBindings(this, widget, m_backend.parameterBindings());

  return widget;
}

bool PythonSource::execute()
{
  resetExecutionFlags();
  resetProgress();
  setExecState(NodeExecState::Running);

  auto outputs = m_backend.runSource(this);

  if (isCanceled()) {
    setExecState(NodeExecState::Canceled);
    return false;
  }

  if (outputs.isEmpty() && !outputPorts().isEmpty()) {
    setExecState(NodeExecState::Failed);
    return false;
  }

  applyOutputs(outputs);

  markCurrent();
  setExecState(NodeExecState::Idle);
  return true;
}

QJsonObject PythonSource::serialize() const
{
  return m_backend.serializeInto(SourceNode::serialize());
}

bool PythonSource::deserialize(const QJsonObject& json)
{
  m_backend.applySerializedFields(
    json,
    nullptr,
    [this](const QString& name, PortType type) {
      return addOutput(name, type);
    });
  setSupportsCancel(m_backend.supportsCancel());
  setSupportsCompletion(m_backend.supportsComplete());
  if (!SourceNode::deserialize(json)) {
    return false;
  }
  if (!nodeExecutor()) {
    auto envPath = m_backend.externalPythonEnvPath();
    if (!envPath.isEmpty()) {
      setNodeExecutor(new ExternalNodeExecutor(envPath));
    }
  }
  return true;
}

} // namespace pipeline
} // namespace tomviz
